src/main/java/com/hadoop/compression/lzo/DistributedLzoIndexer.java

```diff
@@ -25,57 +27,108 @@
   private static final Log LOG = LogFactory.getLog(DistributedLzoIndexer.class);
-  private final String LZO_EXTENSION = new LzopCodec().getDefaultExtension();
 
+  private static final String LZO_EXTENSION = new LzopCodec().getDefaultExtension();
+
+  private static final String LZO_SKIP_INDEXING_SMALL_FILES_KEY = "lzo_skip_indexing_small_files";
+  private static final String LZO_SMALL_FILE_SIZE_KEY = "lzo_small_file_size";
+  private static final String LZO_RECURSIVE_INDEXING_KEY = "lzo_recursive_indexing";
+  private static final boolean LZO_SKIP_INDEXING_SMALL_FILES_DEFAULT = false;
+  private static final boolean LZO_RECURSIVE_INDEXING_DEFAULT = true;
+  private static final long LZO_SMALL_FILE_SIZE_DEFAULT = 0;
+  private boolean lzoSkipIndexingSmallFiles = LZO_SKIP_INDEXING_SMALL_FILES_DEFAULT;
+  private boolean lzoRecursiveIndexing = LZO_RECURSIVE_INDEXING_DEFAULT;
+  private long lzoSmallFileSize = LZO_SMALL_FILE_SIZE_DEFAULT;
+
+  private static final String TEMP_FILE_EXTENSION = "/_temporary";
+
+  private Configuration conf = getConf();
+
+  /**
+   * Accepts paths which don't end in TEMP_FILE_EXTENSION
+   */
   private final PathFilter nonTemporaryFilter = new PathFilter() {
+    @Override
     public boolean accept(Path path) {
-      return !path.toString().endsWith("/_temporary");
+      return !path.toString().endsWith(TEMP_FILE_EXTENSION);
     }
   };
 
-  private void walkPath(Path path, PathFilter pathFilter, List<Path> accumulator) {
+  /**
+   * Returns whether a file should be considered small enough to skip indexing.
+   */
+  private boolean isSmallFile(FileStatus status) {
+    return status.getLen() < lzoSmallFileSize;
+  }
+  
+  private void visitPath(Path path, PathFilter pathFilter, List<Path> accumulator, boolean recursive) {
     try {
-      FileSystem fs = path.getFileSystem(getConf());
+      FileSystem fs = path.getFileSystem(this.conf);
       FileStatus fileStatus = fs.getFileStatus(path);
 
-      if (fileStatus.isDir()) {
-        FileStatus[] children = fs.listStatus(path, pathFilter);
-        for (FileStatus childStatus : children) {
-          walkPath(childStatus.getPath(), pathFilter, accumulator);
-        }
-      } else if (path.toString().endsWith(LZO_EXTENSION)) {
-        Path lzoIndexPath = path.suffix(LzoIndex.LZO_INDEX_SUFFIX);
-        if (fs.exists(lzoIndexPath)) {
-          // If the index exists and is of nonzero size, we're already done.
-          // We re-index a file with a zero-length index, because every file has at least one block.
-          if (fs.getFileStatus(lzoIndexPath).getLen() > 0) {
-            LOG.info("[SKIP] LZO index file already exists for " + path);
-            return;
-          } else {
-            LOG.info("Adding LZO file " + path + " to indexing list (index file exists but is zero length)");
-            accumulator.add(path);
+      if (fileStatus.isDirectory()) {
+        if (recursive) {
```

Comment:
This requires a discussion. This behavior of "recursive" is a little surprising to me. If I'm not mistaken (from things like FileSystem API), recursive=false means *still* processing direct file children of the directory but not traversing into subdirectories. But this patch would skip the directory entirely.

I think it may be good to be consistent and consider direct file children still. Thoughts? Either way, I think we need some comments in the code to clarify what it means.